src/Promise.php

```diff
@@ -2,9 +2,17 @@
 
 namespace React\Promise;
 
+/**
+ * @template-implements PromiseInterface
+ * @template-covariant T
```

Comment:
It would be nice to also have a type parameter for rejection reason. But I have [tried doing that for guzzlehttp/promisees](https://github.com/guzzle/promises/compare/master...jtojnar:guzzle-promises:generic-types-2) and it was a pain.

Reply:
Even if it's a pain, does it work? Guess, based on no PR associated with those changes it doesn't?

One of the issues is literally this: https://github.com/guzzle/promises/compare/master...jtojnar:guzzle-promises:generic-types-2#diff-08c91937f94e1c62ad95bd704218d97cff992fb41a9b71b7fa9519442f8d67dbR33 that checks what is passed in to handle both scenarios not what is coming out of the previous promise.

That is one of the reasons we decided not to bother with rejections (for now).

Reply:
I made a [PR](https://github.com/guzzle/promises/pull/157) but gave up because the various handler propagations were too difficult to reason about and I ran out of time I allocated for the experiment.